modules/packages/pkg

```diff
@@ -0,0 +1,220 @@
+    
+    echo "${INPUTP}" | awk -F= '{print "KEYWORD=" $1 "; VALUE=" $2;}' 2>/dev/null
+
+    unset INPUTP
+
+}
+
+read_options() {
+    while read -r INPUT ; do
+        if [ -n "${INPUT}" ] ; then
+            eval `parse_line "${INPUT}"`
+            case "${KEYWORD}" in
+                File|Name)
+                    VER=''
+                    ARCH=''
+                    if [ "${KEYWORD}" = 'File' ] ; then
+                        PKG="${VALUE}"
+                    else
+                        NAM="${VALUE}"
+                    fi
+                    ;;
+                Version)
+                    VER="${VALUE}"
+                    ;;
+                Architecture)
+                    ARCH=${VALUE:=${DEFARCH}}
+                    ;;
+                options*)
+                    if [ -n "${repository}" ] ; then
+                        REPO="-r ${repository}"
+                    fi
+                    continue
+                    ;;
+                *)
+                    fatal "Invalid input: '${KEYWORD}'."
+                    ;;
+            esac
+            if [ -n "${KEYWORD}" ] && [ -n "${VALUE}" ] ; then
+                INPUTLIST="${INPUTLIST} ${KEYWORD}=${VALUE}"
+            fi
+        fi
+    done
+}
+
+process_inputs() {
+    CMD="$1" ; shift
+    OPTIONS="$*"
+
+    for OPT in ${OPTIONS} ; do
+        eval `parse_line ${OPT}`
+        case "${KEYWORD}" in
+            File*)
+                if [ "${CMD}" = 'file-install' ] ; then
+                    PKG="${PKG} ${VALUE}"
+                else
+                    fatal "Bad Input: '${OPT}' (should be Name=...)."
+                fi
+                VER=''
+                ;;
+            Name*)
+                if [ "${CMD}" != 'file-install' ] ; then
+                    PKG="${PKG} ${VALUE}"
+                else
+                    fatal "Bad Input: '${OPT}' (should be File=...)"
+                fi
+                VER=''
+                ;;
+            Version*)
+                PKG="${PKG}-${VALUE}"
+                ;;
+            Architecture*)
+                if [ -n "${VALUE}" ] &&
+                       [ "${VALUE}" != "${DEFARCH}" ] ; then
+                    fatal "Bad Input: Architecture ${VALUE} does not match ${DEFARCH}."
+                fi
+                ;;
+            *)
+                fatal "Invalid input: '${INPUT}'."
+                ;;
+        esac
+    done
+}
+
+get_package_data() {
+    if [ -n "${PKG}" ] ; then
+        case "${PKG}" in
+            /*|./*|../) # File Path.  It's a file
+                TYP='file'
+                if [ ! -e "${PKG}" ] ; then
+                    fatal "No such file: '${PKG}'."
+                fi
+                ;;
+            http://*|ftp://|file://) # URL, it's a file.
+                TYP='file'
+                ;;
+            */*)
+                if [ -e "${PKG}" ] ; then
+                    TYP='file'
+                else
+                    TYP='repo'
+                fi
```

Comment:
I would not return `file` for a path which isn't absolute. Testing for existence is unpredictable, the appearance of a file might suddenly change CFEngine's idea of what a promise string is (file or repo). I think it's better to test only the "absolute-ness" of the path and that way you'll get a consistent result every time.

This applies to both this `case` entry (`*/*`), and the one above (`/*|./*|../`).